src/librustc_mir/interpret/validity.rs

```diff
@@ -411,11 +411,11 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
                     if !did.is_local() || self.ecx.tcx.is_foreign_item(did) {
                         return Ok(());
                     }
+                    // FIXME: Statics referenced from consts are skipped.
+                    // This avoids spurious "const accesses static" errors
+                    // unrelated to validity, but is similarly unsound.
```

Comment:
I'd merge this with the `if` above, to
```rust
if !did.is_local() || self.ecx.tcx.is_static(did) {
```
The comment should also be updated:
```
// We don't re-validate statics because 
// - the const-machine we are in might forbid reading from them, and
// - they have already been validated.
// FIXME: validation might have happened at a different type, but for now
// we consider this an okay trade-off.
```

Reply:
But your check is stricter in the sense that it validates more, so maybe @oli-obk would prefer that one.

I just think it's generally not worth recursing into other globals -- this can only possibly catch a bug if people did some crazy unsafe stuff, and those people asked for it when they get post-monomorphization errors. So I'd actually be fine to just always bail for any `Some(GlobalAlloc::Static(_))` here.

Reply:
I went with the stricter check because there are tests for this, for statics-pointing-to-statics:
* https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/double_check.rs
* https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/double_check2.rs

Happy to collapse it into the first check if we're okay ditching those tests.

Reply:
Yeah I think it is okay to not catch the failure in double_check2. We already do not catch it cross-crate. But let's wait what @oli-obk thinks -- and in fact let's Cc @rust-lang/wg-const-eval and see what everyone else thinks :D

Reply:
wait, how do we not catch this cross crate? Don't we revisit statics from other crates if we encounter them at a different type of non-zero offset?

Reply:
No we don't, as far as I know.